File: src/y2026/m4_apr/d3_hard/s2751_robot_collisions.rs
// LeetCode Problem 2751. Robot Collisions
// Difficulty: Hard

// Time Complexity: O(m log m) - where m is the size of the positions array
// Space Complexity: O(n) - where n is the size of the positions array

#![allow(dead_code)]
struct Solution;

impl Solution {
    pub fn survived_robots_healths(
        positions: Vec<i32>,
        healths: Vec<i32>,
        directions: String,
    ) -> Vec<i32> {
        let n = positions.len();
        let mut p = (0..n)
            .map(|i| (positions[i], i))
            .collect::<Vec<(i32, usize)>>();
        p.sort();

        let directions = directions.chars().collect::<Vec<char>>();
        let (mut temp, mut sk) = (vec![], vec![]);

        for i in 0..n {
            let k = p[i].1;

            if directions[k] == 'R' {
                sk.push((k, healths[k]));
                continue;
            }

            let mut h = healths[k];

            while !sk.is_empty() {
                let sz = sk.len();
                if sk[sz - 1].1 >= h {
                    break;
                }
                h -= 1;
                sk.pop();
            }

            if !sk.is_empty() {
                let sz = sk.len();
                if sk[sz - 1].1 == h {
                    sk.pop();
                    continue;
                }
                sk[sz - 1].1 -= 1;
                continue;
            }
            temp.push((k, h));
        }
        for (i, h) in sk {
            temp.push((i, h));
        }
        temp.sort();

        temp.into_iter().map(|a| a.1).collect::<Vec<i32>>()
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_survived_robots_healths_1() {
        let result = Solution::survived_robots_healths(
            [5, 4, 3, 2, 1].to_vec(),
            [2, 17, 9, 15, 10].to_vec(),
            "RRRRR".to_string(),
        );

        assert_eq!(result, [2, 17, 9, 15, 10].to_vec())
    }

    #[test]
    fn test_survived_robots_healths_2() {
        let result = Solution::survived_robots_healths(
            [3, 5, 2, 6].to_vec(),
            [10, 10, 15, 12].to_vec(),
            "RLRL".to_string(),
        );

        assert_eq!(result, [14].to_vec())
    }

    #[test]
    fn test_survived_robots_healths_3() {
        let result = Solution::survived_robots_healths(
            [1, 2, 5, 6].to_vec(),
            [10, 10, 11, 11].to_vec(),
            "RLRL".to_string(),
        );

        assert_eq!(result, [].to_vec())
    }
}

File: src/y2026/m4_apr/d3_hard/s2751_robot_collisions.cc
#include <algorithm>
#include <cassert>
#include <cstdlib>
#include <numeric>
#include <stack>
#include <string>
#include <vector>

using namespace std;

class Solution {
 public:
  vector<int> survivedRobotsHealths(vector<int>& positions,
                                    vector<int>& healths, string directions) {
    int n = positions.size();
    vector<int> order(n);
    iota(order.begin(), order.end(), 0);
    sort(order.begin(), order.end(),
         [&](int a, int b) { return positions[a] < positions[b]; });

    stack<int> st;
    vector<bool> dead(n, false);

    for (int i : order) {
      if (directions[i] == 'R') {
        st.push(i);
      } else {
        while (!st.empty() && directions[st.top()] == 'R') {
          int top = st.top();

          if (healths[top] > healths[i]) {
            healths[top]--;
            dead[i] = true;
            break;
          } else if (healths[top] < healths[i]) {
            healths[i]--;
            dead[top] = true;
            st.pop();
          } else {
            dead[i] = true;
            dead[top] = true;
            st.pop();
            break;
          }
        }
        if (!dead[i]) {
          st.push(i);
        }
      }
    }

    vector<int> result;
    for (int i = 0; i < n; i++) {
      if (!dead[i]) {
        result.push_back(healths[i]);
      }
    }

    return result;
  }
};

int main() {
  Solution sol;

  // Example 1
  {
    vector<int> positions = {5, 4, 3, 2, 1};
    vector<int> healths = {2, 17, 9, 15, 10};
    string directions = "RRRRR";
    vector<int> result =
        sol.survivedRobotsHealths(positions, healths, directions);
    vector<int> expected = {2, 17, 9, 15, 10};
    assert(result == expected);
  }

  // Example 2
  {
    vector<int> positions = {3, 5, 2, 6};
    vector<int> healths = {10, 10, 15, 12};
    string directions = "RLRL";
    vector<int> result =
        sol.survivedRobotsHealths(positions, healths, directions);
    vector<int> expected = {14};
    assert(result == expected);
  }

  // Example 3
  {
    vector<int> positions = {1, 2, 5, 6};
    vector<int> healths = {10, 10, 11, 11};
    string directions = "RLRL";
    vector<int> result =
        sol.survivedRobotsHealths(positions, healths, directions);
    vector<int> expected = {};
    assert(result == expected);
  }

  return EXIT_SUCCESS;
}
